jerry-ext/handle-scope/handle-scope.c

```diff
@@ -0,0 +1,370 @@
+  return jerryx_handle_scope_ok;
+} /** jerryx_escape_handle_internal */
+
+
+/**
+ * Promotes the handle to the JavaScript object so that it is valid for the lifetime of
+ * the outer scope. It can only be called once per scope. If it is called more than
+ * once an error will be returned.
+ *
+ * @param scope - the expected scope to be escaped from.
+ * @param escapee - the jerry value to be escaped.
+ * @param result - [out value] escaped jerry value result.
+ * @return status code, jerryx_handle_scope_ok if success.
+ */
+jerryx_handle_scope_status
+jerryx_escape_handle (jerryx_escapable_handle_scope scope,
+                      jerry_value_t escapee,
+                      jerry_value_t *result)
+{
+  return jerryx_escape_handle_internal (scope, escapee, result, true);
+} /** jerryx_escape_handle */
+
+
+/**
+ * Escape a handle from scope yet do not promote it to the outer scope.
+ * Leave the handle's life time management up to user.
+ *
+ * @param scope - the expected scope to be removed from.
+ * @param escapee - the jerry value to be removed.
+ * @param result - [out value] removed jerry value result.
+ * @return status code, jerryx_handle_scope_ok if success.
+ */
+jerryx_handle_scope_status
+jerryx_remove_handle (jerryx_escapable_handle_scope scope,
+                      jerry_value_t escapee,
+                      jerry_value_t *result)
+{
+  return jerryx_escape_handle_internal (scope, escapee, result, false);
+} /** jerryx_remove_handle */
+
+
+/**
+ * Try to reuse given handle if possible while adding to the scope.
+ *
+ * @param handle - the one to be added to the scope.
+ * @param scope - the scope of handle to be added to.
+ * @returns the jerry value id wrapped by given handle.
+ */
+jerry_value_t
+jerryx_handle_scope_add_handle_to (jerryx_handle_t *handle, jerryx_handle_scope scope)
+{
+  size_t handle_count = scope->handle_count;
+  scope->handle_count += 1;
+  if (handle_count < JERRYX_HANDLE_PRELIST_SIZE)
+  {
+    jerry_value_t jval = handle->jval;
+    free (handle);
+    scope->handle_prelist[handle_count] = jval;
+    return jval;
+  }
+
+  handle->sibling = scope->handle_ptr;
+  scope->handle_ptr = handle;
+  return handle->jval;
+} /** jerryx_handle_scope_add_handle_to */
+
+
+/**
+ * Add given jerry value to the scope.
+ *
+ * @param jval - jerry value to be added to scope.
+ * @param scope - the scope of the jerry value been expected to be added to.
+ * @return jerry value that added to scope.
+ */
+jerry_value_t
+jerryx_create_handle_in_scope (jerry_value_t jval, jerryx_handle_scope scope)
+{
+  size_t handle_count = scope->handle_count;
+  if (handle_count < JERRYX_HANDLE_PRELIST_SIZE)
+  {
+    scope->handle_prelist[handle_count] = jval;
+    goto deferred;
+  }
+  jerryx_handle_t *handle = malloc (sizeof (jerryx_handle_t));
+  JERRYX_HANDLE_SCOPE_ASSERT (handle != NULL);
+  handle->jval = jval;
+
+  handle->sibling = scope->handle_ptr;
+  scope->handle_ptr = handle;
+
+deferred:
+  scope->handle_count += 1;
```

Comment:
Is there any reason of using 
`+= 1 ` and `-= 1` everywhere, instead of `++/--` ?